crates/bevy_render/src/render_resource/bind_group_layout_entries.rs

```diff
@@ -0,0 +1,582 @@
+                .into_array()
+                .into_iter()
+                .enumerate()
+                .map(|(ix, resource)| resource.build(ix as u32, default_visibility))
+                .collect(),
+        }
+    }
+
+    pub fn extend_sequential<const N: usize>(
+        mut self,
+        entries: impl IntoBindGroupLayoutEntryBuilderArray<N>,
+    ) -> Self {
+        let start = self.entries.last().unwrap().binding + 1;
+        self.entries.extend(
+            entries
+                .into_array()
+                .into_iter()
+                .enumerate()
+                .map(|(ix, resource)| resource.build(start + ix as u32, self.default_visibility)),
+        );
+        self
+    }
+
+    pub fn new_with_indices<const N: usize>(
+        default_visibility: ShaderStages,
+        entries: impl IntoIndexedBindGroupLayoutEntryBuilderArray<N>,
+    ) -> Self {
+        Self {
+            default_visibility,
+            entries: entries
+                .into_array()
+                .into_iter()
+                .map(|(binding, resource)| resource.build(binding, default_visibility))
+                .collect(),
+        }
+    }
+
+    pub fn extend_with_indices<const N: usize>(
+        mut self,
+        entries: impl IntoIndexedBindGroupLayoutEntryBuilderArray<N>,
+    ) -> Self {
+        self.entries.extend(
+            entries
+                .into_array()
+                .into_iter()
+                .map(|(binding, resource)| resource.build(binding, self.default_visibility)),
+        );
+        self
+    }
+}
+
+impl std::ops::Deref for DynamicBindGroupLayoutEntries {
+    type Target = [BindGroupLayoutEntry];
+
+    fn deref(&self) -> &[BindGroupLayoutEntry] {
+        &self.entries
+    }
+}
+
+pub mod binding_types {
+    use crate::render_resource::{
+        BufferBindingType, SamplerBindingType, TextureSampleType, TextureViewDimension,
+    };
+    use std::num::NonZeroU64;
+    use wgpu::{BindingType, StorageTextureAccess, TextureFormat};
+
+    use super::*;
+
+    #[allow(unused)]
+    pub fn storage_buffer(
+        has_dynamic_offset: bool,
+        min_binding_size: Option<NonZeroU64>,
+    ) -> BindGroupLayoutEntryBuilder {
+        BindingType::Buffer {
+            ty: BufferBindingType::Storage { read_only: false },
+            has_dynamic_offset,
+            min_binding_size,
+        }
+        .into_bind_group_layout_entry_builder()
+    }
+
+    #[allow(unused)]
+    pub fn storage_buffer_read_only(
+        has_dynamic_offset: bool,
+        min_binding_size: Option<NonZeroU64>,
+    ) -> BindGroupLayoutEntryBuilder {
+        BindingType::Buffer {
+            ty: BufferBindingType::Storage { read_only: true },
+            has_dynamic_offset,
+            min_binding_size,
+        }
+        .into_bind_group_layout_entry_builder()
+    }
+
+    #[allow(unused)]
+    pub fn uniform_buffer(
+        has_dynamic_offset: bool,
+        min_binding_size: Option<NonZeroU64>,
```

Comment:
since nearly every in-engine use provides the size, perhaps we could make this `uniform_buffer<T: ShaderType>` and do the min size automatically?

we would probably want a `uniform_buffer_sized` variant available as well where you can pass it manually.

Reply:
I like this idea, but what about the case where size is None?

Reply:
Yeah, I tried it and it's really nice. I also implemented it for storage buffers